src/main/java/org/opensearch/security/privileges/PrivilegesEvaluatorResponse.java

```diff
@@ -26,33 +26,65 @@
 
 package org.opensearch.security.privileges;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
 
+import com.google.common.collect.ImmutableSet;
+
 import org.opensearch.action.admin.indices.create.CreateIndexRequestBuilder;
 import org.opensearch.security.resolver.IndexResolverReplacer.Resolved;
 import org.opensearch.security.securityconf.EvaluatedDlsFlsConfig;
 
 public class PrivilegesEvaluatorResponse {
     boolean allowed = false;
-    Set<String> missingPrivileges = new HashSet<String>();
     Set<String> missingSecurityRoles = new HashSet<>();
     Set<String> resolvedSecurityRoles = new HashSet<>();
     EvaluatedDlsFlsConfig evaluatedDlsFlsConfig;
     PrivilegesEvaluatorResponseState state = PrivilegesEvaluatorResponseState.PENDING;
     Resolved resolved;
     CreateIndexRequestBuilder createIndexRequestBuilder;
+    private ImmutableSet<String> onlyAllowedForIndices = ImmutableSet.of();
+    private CheckTable<String, String> indexToActionCheckTable;
+    private String reason;
 
     public Resolved getResolved() {
         return resolved;
     }
 
+    /**
+     * Returns true if the request can be fully allowed. See also isAllowedForSpecificIndices().
+     */
     public boolean isAllowed() {
         return allowed;
     }
 
+    /**
+     * Returns true if the request can be allowed if the referenced indices are reduced (aka "do not fail on forbidden")
+     */
+    public boolean isPartiallyOk() {
+        return !this.onlyAllowedForIndices.isEmpty();
+    }
+
+    public ImmutableSet<String> getAvailableIndices() {
+        return this.onlyAllowedForIndices;
+    }
+
     public Set<String> getMissingPrivileges() {
-        return new HashSet<String>(missingPrivileges);
+        return this.indexToActionCheckTable != null ? this.indexToActionCheckTable.getIncompleteColumns() : Collections.emptySet();
+    }
+
+    public String getReason() {
+        return this.reason;
+    }
+
+    public PrivilegesEvaluatorResponse reason(String reason) {
+        this.reason = reason;
+        return this;
+    }
+
+    public CheckTable<String, String> getCheckTable() {
```

Comment:
Lets not bleed this data type outside of the class / isn't this table another form of the 'reason' or the toString?